File: Scheduler.cpp
/*
 * Scheduler.cpp
 * Implementation of Scheduler class
 */

#include "Scheduler.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <iomanip>
#include <queue>

Scheduler::Scheduler() : timeQuantum(1.0)
{
}

bool Scheduler::readInputFile(const std::string &filename)
{
    std::ifstream file(filename);
    if (!file.is_open())
    {
        std::cerr << "Error: Could not open file '" << filename << "'" << std::endl;
        return false;
    }

    processes.clear();
    std::string line;
    int processID = 1;

    // Read first line for time quantum
    if (std::getline(file, line))
    {
        std::stringstream ss(line);
        std::string token;
        if (std::getline(ss, token, ','))
        {
            timeQuantum = std::stod(token);
        }
    }

    // Read process data
    while (std::getline(file, line))
    {
        if (line.empty())
            continue;

        std::stringstream ss(line);
        std::string token;

        while (std::getline(ss, token, ','))
        {
            double burstTime = std::stod(token);
            if (burstTime > 0)
            {
                Process p(processID++, 0.0, burstTime, 0);
                processes.push_back(p);
            }
        }
    }

    file.close();
    std::cout << "Successfully loaded " << processes.size() << " processes" << std::endl;
    return true;
}

void Scheduler::displayProcesses() const
{
    std::cout << "\n=== Loaded Processes ===" << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << std::setw(4) << "PID"
              << std::setw(8) << "AT"
              << std::setw(8) << "BT" << std::endl;
    std::cout << std::string(20, '-') << std::endl;

    for (const auto &p : processes)
    {
        std::cout << std::setw(4) << p.getPID()
                  << std::setw(8) << p.getArrivalTime()
                  << std::setw(8) << p.getBurstTime() << std::endl;
    }
}

void Scheduler::displayHeader() const
{
    std::cout << std::fixed << std::setprecision(2);
    std::cout << std::setw(4) << "PID"
              << std::setw(8) << "AT"
              << std::setw(8) << "BT"
              << std::setw(8) << "CT"
              << std::setw(8) << "WT"
              << std::setw(10) << "TAT"
              << std::setw(10) << "RT" << std::endl;
    std::cout << std::string(54, '-') << std::endl;
}

void Scheduler::displayResults(const std::vector<Process> &procs, const std::string &algorithm) const
{
    std::cout << "\n=== " << algorithm << " Results ===" << std::endl;
    displayHeader();

    double totalWT = 0, totalTAT = 0, totalRT = 0;

    for (const auto &p : procs)
    {
        p.display();
        totalWT += p.getWaitingTime();
        totalTAT += p.getTurnaroundTime();
        totalRT += p.getResponseTime();
    }

    int n = procs.size();
    std::cout << std::string(54, '-') << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Average Waiting Time: " << totalWT / n << std::endl;
    std::cout << "Average Turnaround Time: " << totalTAT / n << std::endl;
    std::cout << "Average Response Time: " << totalRT / n << std::endl;
}

void Scheduler::resetProcesses()
{
    for (auto &p : processes)
    {
        p.reset();
    }
}

// ========== SCHEDULING ALGORITHMS ==========

std::vector<Process> Scheduler::fcfsScheduling()
{
    std::cout << "\n=== First Come First Serve (FCFS) Scheduling ===" << std::endl;
    resetProcesses();

    std::vector<Process> scheduled = processes;

    // Sort by arrival time
    std::sort(scheduled.begin(), scheduled.end(),
              [](const Process &a, const Process &b)
              {
                  return a.getArrivalTime() < b.getArrivalTime();
              });

    double currentTime = 0.0;

    // FCFS: Execute processes in order of arrival
    for (auto &p : scheduled)
    {
        // If CPU is idle, wait for the process to arrive
        if (currentTime < p.getArrivalTime())
        {
            currentTime = p.getArrivalTime();
        }

        // Response time = time when process first gets CPU - arrival time
        double responseTime = currentTime - p.getArrivalTime();
        p.setResponseTime(responseTime);

        // Execute the process
        currentTime += p.getBurstTime();

        // Completion time = time when process finishes
        p.setCompletionTime(currentTime);

        // Turnaround time = completion time - arrival time
        double turnaroundTime = currentTime - p.getArrivalTime();
        p.setTurnaroundTime(turnaroundTime);

        // Waiting time = turnaround time - burst time
        double waitingTime = turnaroundTime - p.getBurstTime();
        p.setWaitingTime(waitingTime);
    }

    displayResults(scheduled, "FCFS");
    return scheduled;
}

std::vector<Process> Scheduler::sjfScheduling()
{
    std::cout << "\n=== Shortest Job First (SJF) Scheduling ===" << std::endl;
    resetProcesses();

    std::vector<Process> scheduled = processes;

    // TODO: Implement SJF logic
    // Sort by burst time (shortest first)
    // Then apply similar timing calculations as FCFS

    std::sort(scheduled.begin(), scheduled.end(),
              [](const Process &a, const Process &b)
              {
                  return a.getBurstTime() < b.getBurstTime();
              });

    double currentTime = 0.0;

    // SJF: Execute processes in order of shortest burst time
    for (auto &p : scheduled)
    {
        // If CPU is idle, wait for the process to arrive
        if (currentTime < p.getArrivalTime())
        {
            currentTime = p.getArrivalTime();
        }

        // Response time = time when process first gets CPU - arrival time
        double responseTime = currentTime - p.getArrivalTime();
        p.setResponseTime(responseTime);

        // Execute the process
        currentTime += p.getBurstTime();

        // Completion time = time when process finishes
        p.setCompletionTime(currentTime);

        // Turnaround time = completion time - arrival time
        double turnaroundTime = currentTime - p.getArrivalTime();
        p.setTurnaroundTime(turnaroundTime);

        // Waiting time = turnaround time - burst time
        double waitingTime = turnaroundTime - p.getBurstTime();
        p.setWaitingTime(waitingTime);
    }

    displayResults(scheduled, "SJF");
    return scheduled;
}

std::vector<Process> Scheduler::priorityScheduling()
{
    std::cout << "\n=== Priority Scheduling ===" << std::endl;
    resetProcesses();

    std::vector<Process> scheduled = processes;

    // Sort by priority (lower number = higher priority)
    std::sort(scheduled.begin(), scheduled.end(),
              [](const Process &a, const Process &b)
              {
                  return a.getPriority() < b.getPriority();
              });

    double currentTime = 0.0;

    for (auto &p : scheduled)
    {
        // If current time is less than arrival time, CPU is idle
        if (currentTime < p.getArrivalTime())
        {
            currentTime = p.getArrivalTime();
        }

        // Response time = time when process starts - arrival time
        double responseTime = currentTime - p.getArrivalTime();
        p.setResponseTime(responseTime);

        // Process executes for its burst time
        currentTime += p.getBurstTime();

        // Completion time = time when process finishes
        p.setCompletionTime(currentTime);

        // Turnaround time = completion time - arrival time
        double turnaroundTime = currentTime - p.getArrivalTime();
        p.setTurnaroundTime(turnaroundTime);

        // Waiting time = turnaround time - burst time
        double waitingTime = turnaroundTime - p.getBurstTime();
        p.setWaitingTime(waitingTime);
    }

    displayResults(scheduled, "Priority");
    return scheduled;
}

std::vector<Process> Scheduler::roundRobinScheduling()
{
    std::cout << "\n=== Round Robin Scheduling (TQ: " << timeQuantum << ") ===" << std::endl;
    resetProcesses();

    std::vector<Process> scheduled = processes;
    std::queue<int> readyQueue;
    std::vector<bool> inQueue(scheduled.size(), false);
    std::vector<bool> firstTime(scheduled.size(), true);
    double currentTime = 0.0;
    int completed = 0;
    int n = scheduled.size();

    // Add processes that have arrived at time 0
    for (size_t i = 0; i < scheduled.size(); i++)
    {
        if (scheduled[i].getArrivalTime() <= currentTime)
        {
            readyQueue.push(i);
            inQueue[i] = true;
        }
    }

    while (completed < n)
    {
        if (readyQueue.empty())
        {
            // Find next process to arrive
            double nextArrival = 1e9;
            for (size_t i = 0; i < scheduled.size(); i++)
            {
                if (scheduled[i].getRemainingTime() > 0 && !inQueue[i])
                {
                    nextArrival = std::min(nextArrival, scheduled[i].getArrivalTime());
                }
            }
            currentTime = nextArrival;

            // Add newly arrived processes
            for (size_t i = 0; i < scheduled.size(); i++)
            {
                if (scheduled[i].getArrivalTime() <= currentTime &&
                    scheduled[i].getRemainingTime() > 0 && !inQueue[i])
                {
                    readyQueue.push(i);
                    inQueue[i] = true;
                }
            }
            continue;
        }

        int idx = readyQueue.front();
        readyQueue.pop();
        inQueue[idx] = false;

        // Set response time when process first gets CPU
        if (firstTime[idx])
        {
            double responseTime = currentTime - scheduled[idx].getArrivalTime();
            scheduled[idx].setResponseTime(responseTime);
            firstTime[idx] = false;
        }

        // Execute for time quantum or remaining time (whichever is smaller)
        double execTime = std::min(timeQuantum, scheduled[idx].getRemainingTime());
        scheduled[idx].setRemainingTime(scheduled[idx].getRemainingTime() - execTime);
        currentTime += execTime;

        // Check for newly arrived processes during execution
        for (size_t i = 0; i < scheduled.size(); i++)
        {
            if (i != (size_t)idx && scheduled[i].getArrivalTime() <= currentTime &&
                scheduled[i].getRemainingTime() > 0 && !inQueue[i])
            {
                readyQueue.push(i);
                inQueue[i] = true;
            }
        }

        // If process is complete
        if (scheduled[idx].getRemainingTime() == 0)
        {
            completed++;
            scheduled[idx].setCompletionTime(currentTime);
            double turnaroundTime = currentTime - scheduled[idx].getArrivalTime();
            scheduled[idx].setTurnaroundTime(turnaroundTime);
            double waitingTime = turnaroundTime - scheduled[idx].getBurstTime();
            scheduled[idx].setWaitingTime(waitingTime);
        }
        else
        {
            // Process not finished, put back in queue
            readyQueue.push(idx);
            inQueue[idx] = true;
        }
    }

    displayResults(scheduled, "Round Robin");
    return scheduled;
}

std::vector<Process> Scheduler::srtfScheduling()
{
    std::cout << "\n=== Shortest Remaining Time First (SRTF) Scheduling ===" << std::endl;
    resetProcesses();

    std::vector<Process> scheduled = processes;
    double currentTime = 0.0;
    int completed = 0;
    int n = scheduled.size();
    std::vector<bool> firstTime(n, true);
    int lastProcess = -1;

    while (completed < n)
    {
        // Find process with shortest remaining time among arrived processes
        int shortest = -1;
        double minRemaining = 1e9;

        for (int i = 0; i < n; i++)
        {
            if (scheduled[i].getArrivalTime() <= currentTime &&
                scheduled[i].getRemainingTime() > 0 &&
                scheduled[i].getRemainingTime() < minRemaining)
            {
                minRemaining = scheduled[i].getRemainingTime();
                shortest = i;
            }
        }

        // If no process available, jump to next arrival
        if (shortest == -1)
        {
            double nextArrival = 1e9;
            for (int i = 0; i < n; i++)
            {
                if (scheduled[i].getRemainingTime() > 0)
                {
                    nextArrival = std::min(nextArrival, scheduled[i].getArrivalTime());
                }
            }
            currentTime = nextArrival;
            continue;
        }

        // Set response time when process first gets CPU
        if (firstTime[shortest])
        {
            double responseTime = currentTime - scheduled[shortest].getArrivalTime();
            scheduled[shortest].setResponseTime(responseTime);
            firstTime[shortest] = false;
        }

        // Execute for a small time unit (0.01 for precision)
        double execTime = 0.01;
        if (scheduled[shortest].getRemainingTime() < execTime)
        {
            execTime = scheduled[shortest].getRemainingTime();
        }

        scheduled[shortest].setRemainingTime(scheduled[shortest].getRemainingTime() - execTime);
        currentTime += execTime;

        // If process completed
        if (scheduled[shortest].getRemainingTime() < 0.0001)
        { // Using small epsilon for floating point comparison
            completed++;
            scheduled[shortest].setCompletionTime(currentTime);
            double turnaroundTime = currentTime - scheduled[shortest].getArrivalTime();
            scheduled[shortest].setTurnaroundTime(turnaroundTime);
            double waitingTime = turnaroundTime - scheduled[shortest].getBurstTime();
            scheduled[shortest].setWaitingTime(waitingTime);
            scheduled[shortest].setRemainingTime(0); // Ensure it's exactly 0
        }

        lastProcess = shortest;
    }

    displayResults(scheduled, "SRTF");
    return scheduled;
}

void Scheduler::runAllAlgorithms()
{
    std::cout << "\n"
              << std::string(60, '=') << std::endl;
    std::cout << "Running All Scheduling Algorithms" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    fcfsScheduling();
    sjfScheduling();
    priorityScheduling();
    roundRobinScheduling();
    srtfScheduling();
}

void Scheduler::displayMenu()
{
    std::cout << "\n"
              << std::string(50, '=') << std::endl;
    std::cout << "CPU Scheduling Simulator" << std::endl;
    std::cout << std::string(50, '=') << std::endl;
    std::cout << "1. First Come First Serve (FCFS)" << std::endl;
    std::cout << "2. Shortest Job First (SJF)" << std::endl;
    std::cout << "3. Priority Scheduling" << std::endl;
    std::cout << "4. Round Robin (RR)" << std::endl;
    std::cout << "5. Shortest Remaining Time First (SRTF)" << std::endl;
    std::cout << "6. Run All Algorithms" << std::endl;
    std::cout << "7. Display Processes" << std::endl;
    std::cout << "8. Exit" << std::endl;
    std::cout << "\nEnter your choice (1-8): ";
}

File: scheduler.py
"""
Operating System Project - CPU Scheduling Algorithms
Author: [Your Name]
Date: November 2025
"""

import sys
from typing import List, Tuple

class Process:
    """Represents a process with its properties"""
    def __init__(self, pid: int, arrival_time: float = 0, burst_time: float = 0, priority: int = 0):
        self.pid = pid
        self.arrival_time = arrival_time
        self.burst_time = burst_time
        self.priority = priority
        self.waiting_time = 0
        self.turnaround_time = 0
        self.completion_time = 0
        self.response_time = 0
        self.remaining_time = burst_time

    def __str__(self):
        return f"P{self.pid}: AT={self.arrival_time}, BT={self.burst_time}, WT={self.waiting_time:.2f}, TAT={self.turnaround_time:.2f}"

class CPUScheduler:
    """Main scheduler class implementing various scheduling algorithms"""
    
    def __init__(self):
        self.processes = []
        self.time_quantum = 0
    
    def read_input(self, filename: str = "input.txt") -> bool:
        """
        Read process data from input file
        Expected format:
        - First line: time_quantum, [other parameters]
        - Following lines: process burst times or arrival_time, burst_time pairs
        """
        try:
            with open(filename, 'r') as file:
                lines = file.readlines()
            
            # Parse first line for parameters
            first_line = [float(x.strip()) for x in lines[0].split(',')]
            self.time_quantum = first_line[0] if len(first_line) > 0 else 2.0
            
            # Parse process data
            self.processes = []
            process_id = 1
            
            for i, line in enumerate(lines[1:], 1):
                if line.strip():  # Skip empty lines
                    times = [float(x.strip()) for x in line.split(',') if x.strip()]
                    
                    # Create processes from the line data
                    for j, burst_time in enumerate(times):
                        if burst_time > 0:  # Only add valid processes
                            process = Process(
                                pid=process_id,
                                arrival_time=0,  # Default arrival time
                                burst_time=burst_time
                            )
                            self.processes.append(process)
                            process_id += 1
            
            print(f"Successfully loaded {len(self.processes)} processes")
            return True
            
        except FileNotFoundError:
            print(f"Error: File '{filename}' not found!")
            return False
        except Exception as e:
            print(f"Error reading file: {e}")
            return False
    
    def reset_processes(self):
        """Reset all process timing data for new algorithm execution"""
        for process in self.processes:
            process.waiting_time = 0
            process.turnaround_time = 0
            process.completion_time = 0
            process.response_time = 0
            process.remaining_time = process.burst_time
    
    def fcfs_scheduling(self) -> List[Process]:
        """First Come First Serve scheduling algorithm"""
        print("\n=== First Come First Serve (FCFS) Scheduling ===")
        self.reset_processes()
        
        # Sort by arrival time
        scheduled_processes = sorted(self.processes, key=lambda p: p.arrival_time)
        
        current_time = 0
        for process in scheduled_processes:
            # TODO: Implement FCFS logic
            # Calculate waiting time, turnaround time, completion time
            pass
        
        return scheduled_processes
    
    def sjf_scheduling(self) -> List[Process]:
        """Shortest Job First scheduling algorithm"""
        print("\n=== Shortest Job First (SJF) Scheduling ===")
        self.reset_processes()
        
        # TODO: Implement SJF logic
        # Sort by burst time
        scheduled_processes = sorted(self.processes, key=lambda p: p.burst_time)
        
        return scheduled_processes
    
    def priority_scheduling(self) -> List[Process]:
        """Priority scheduling algorithm"""
        print("\n=== Priority Scheduling ===")
        self.reset_processes()
        
        # TODO: Implement Priority scheduling logic
        # Sort by priority (assuming lower number = higher priority)
        scheduled_processes = sorted(self.processes, key=lambda p: p.priority)
        
        return scheduled_processes
    
    def round_robin_scheduling(self) -> List[Process]:
        """Round Robin scheduling algorithm"""
        print(f"\n=== Round Robin Scheduling (Time Quantum: {self.time_quantum}) ===")
        self.reset_processes()
        
        # TODO: Implement Round Robin logic
        scheduled_processes = self.processes.copy()
        
        return scheduled_processes
    
    def srtf_scheduling(self) -> List[Process]:
        """Shortest Remaining Time First scheduling algorithm"""
        print("\n=== Shortest Remaining Time First (SRTF) Scheduling ===")
        self.reset_processes()
        
        # TODO: Implement SRTF logic (preemptive SJF)
        scheduled_processes = self.processes.copy()
        
        return scheduled_processes
    
    def calculate_metrics(self, processes: List[Process]) -> Tuple[float, float, float]:
        """Calculate average waiting time, turnaround time, and response time"""
        if not processes:
            return 0.0, 0.0, 0.0
        
        total_wt = sum(p.waiting_time for p in processes)
        total_tat = sum(p.turnaround_time for p in processes)
        total_rt = sum(p.response_time for p in processes)
        
        n = len(processes)
        avg_wt = total_wt / n
        avg_tat = total_tat / n
        avg_rt = total_rt / n
        
        return avg_wt, avg_tat, avg_rt
    
    def display_results(self, processes: List[Process], algorithm_name: str):
        """Display scheduling results in a formatted table"""
        print(f"\n{algorithm_name} Results:")
        print("-" * 80)
        print(f"{'PID':<4} {'AT':<6} {'BT':<6} {'CT':<6} {'WT':<6} {'TAT':<6} {'RT':<6}")
        print("-" * 80)
        
        for process in processes:
            print(f"{process.pid:<4} {process.arrival_time:<6.1f} {process.burst_time:<6.1f} "
                  f"{process.completion_time:<6.1f} {process.waiting_time:<6.1f} "
                  f"{process.turnaround_time:<6.1f} {process.response_time:<6.1f}")
        
        avg_wt, avg_tat, avg_rt = self.calculate_metrics(processes)
        print("-" * 80)
        print(f"Average Waiting Time: {avg_wt:.2f}")
        print(f"Average Turnaround Time: {avg_tat:.2f}")
        print(f"Average Response Time: {avg_rt:.2f}")
    
    def run_all_algorithms(self):
        """Run all scheduling algorithms and display results"""
        algorithms = [
            ("FCFS", self.fcfs_scheduling),
            ("SJF", self.sjf_scheduling),
            ("Priority", self.priority_scheduling),
            ("Round Robin", self.round_robin_scheduling),
            ("SRTF", self.srtf_scheduling)
        ]
        
        print("Running all scheduling algorithms...")
        print("=" * 50)
        
        for name, algorithm in algorithms:
            try:
                result = algorithm()
                self.display_results(result, name)
                print()
            except Exception as e:
                print(f"Error in {name} algorithm: {e}")

def main():
    """Main function to run the scheduler"""
    scheduler = CPUScheduler()
    
    # Read input file
    if not scheduler.read_input():
        sys.exit(1)
    
    # Display loaded processes
    print("\nLoaded Processes:")
    for process in scheduler.processes:
        print(process)
    
    # Run scheduling algorithms
    while True:
        print("\n" + "="*50)
        print("CPU Scheduling Simulator")
        print("="*50)
        print("1. First Come First Serve (FCFS)")
        print("2. Shortest Job First (SJF)")
        print("3. Priority Scheduling")
        print("4. Round Robin (RR)")
        print("5. Shortest Remaining Time First (SRTF)")
        print("6. Run All Algorithms")
        print("7. Reload Input File")
        print("8. Exit")
        
        choice = input("\nEnter your choice (1-8): ").strip()
        
        if choice == '1':
            result = scheduler.fcfs_scheduling()
            scheduler.display_results(result, "FCFS")
        elif choice == '2':
            result = scheduler.sjf_scheduling()
            scheduler.display_results(result, "SJF")
        elif choice == '3':
            result = scheduler.priority_scheduling()
            scheduler.display_results(result, "Priority")
        elif choice == '4':
            result = scheduler.round_robin_scheduling()
            scheduler.display_results(result, "Round Robin")
        elif choice == '5':
            result = scheduler.srtf_scheduling()
            scheduler.display_results(result, "SRTF")
        elif choice == '6':
            scheduler.run_all_algorithms()
        elif choice == '7':
            scheduler.read_input()
        elif choice == '8':
            print("Exiting... Goodbye!")
            break
        else:
            print("Invalid choice! Please enter a number between 1-8.")

if __name__ == "__main__":
    main()